### PR TITLE
Data driven process container

### PR DESCRIPTION
**Pull request by lielShraga**

### CODE REVIEW

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 1]
```diff
@@ -27,7 +27,9 @@ import akka.stream.scaladsl.{Flow, Framing, Keep, Sink, Source}
 import akka.util.{ByteString, ByteStringBuilder}
 import cmwell.dc.{LazyLogging, Settings}
 import cmwell.dc.Settings._
+import cmwell.dc.stream.InfotonAllMachinesDistributerAndIngester.logger
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: ???

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 2]
```diff
@@ -27,7 +27,9 @@ import akka.stream.scaladsl.{Flow, Framing, Keep, Sink, Source}
...
 import cmwell.dc.stream.MessagesTypesAndExceptions._
+import cmwell.dc.stream.SingleMachineInfotonIngester.IngestStateStatus
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: This type shouldn't be here

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 3]
```diff
@@ -243,6 +245,39 @@ object InfotonRetriever extends LazyLogging {
   ): (Try[RetrieveOutput], RetrieveState) =
     response match {
       case (response @ Success(res), state @ (input, status), body) => {
+        res.collect {
+          case (id, _) if id.data == empty => id
+        }
+//        if (!originalRequest.contains("meta/sys#indexTime")) {
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: delete

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 4]
```diff
@@ -243,6 +245,39 @@ object InfotonRetriever extends LazyLogging {
   ): (Try[RetrieveOutput], RetrieveState) =
     response match {
       case (response @ Success(res), state @ (input, status), body) => {
+        res.collect {
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: dead code

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 5]
```diff
@@ -251,7 +286,33 @@ object InfotonRetriever extends LazyLogging {
           case (id, casIndexTime) if casIndexTime.isDefined && id.meta.indexTime != casIndexTime.get =>
             id
         }
-        if (missingUuids.isEmpty && uuidsWithBadIndexTime.isEmpty) {
+        val uuidsWithMissingIndextime = res.collect{
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: comment

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 6]
```diff
@@ -251,7 +286,33 @@ object InfotonRetriever extends LazyLogging {
...
+          case(id, None) => id
+        }
+
+        if(uuidsWithMissingIndextime.nonEmpty){
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: should be refactored

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 7]
```diff
@@ -15,35 +15,29 @@
...
-import akka.actor.{
-  Actor,
-  ActorSystem,
-  Cancellable,
-  OneForOneStrategy,
-  Props,
-  Status
-}
-import akka.event.{BusLogging, Logging}
+import akka.actor.{Actor, ActorSystem, Cancellable, OneForOneStrategy, Props, Status}
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: the file will be reviewed during the next phase

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonAggregator.scala [hunk 8]
```diff
@@ -19,7 +19,7 @@ import akka.stream.scaladsl.Flow
 import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
 import akka.util.ByteString
 import cmwell.dc.stream.InfotonAggregator.InfotonBucket
-import cmwell.dc.stream.MessagesTypesAndExceptions.{InfotonData, InfotonMeta}
+import cmwell.dc.stream.MessagesTypesAndExceptions.{InfotonData, InfotonFullMeta}
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonAggregator.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala [hunk 9]
```diff
@@ -45,7 +50,21 @@ object MessagesTypesAndExceptions {
   case object CheckDcInfotonList
   case class RetrievedDcInfoList(dcInfoSeq: Seq[DcInfo])
 
-  case class InfotonMeta(path: String, uuid: ByteString, indexTime: Long)
+  sealed trait InfotonMeta{
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala]: done

File: server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package ddpc.data
+
+import java.net.{URL, URLClassLoader}
+
+import akka.actor.ActorSystem
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.{Framing, Source}
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala]: done

File: server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import akka.util.ByteString
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala]: done

File: server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  *
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package ddpc.data
+
+import java.net.{URL, URLClassLoader}
+
+import akka.actor.ActorSystem
+import akka.stream.ActorMaterializer
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala]: done

File: server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  * an “AS IS” BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  *
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package ddpc.data
+
+import java.net.{URL, URLClassLoader}
+
+import akka.actor.ActorSystem
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala]: done

File: server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.{Framing, Source}
+import akka.util.ByteString
+import ChildFirstURLClassLoader.Loader
+
+import scala.concurrent.ExecutionContext
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala]: done

File: server/cmwell-util/src/main/scala/cmwell/util/loading/UrlClassLoader.scala [hunk 15]
```diff
@@ -0,0 +1,37 @@
+/**
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-util/src/main/scala/cmwell/util/loading/UrlClassLoader.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala (hunk 1 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: done.removed

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala (hunk 2 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: removed.done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala (hunk 4 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: removed.done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala (hunk 3 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala (hunk 5 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala (hunk 6 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/Util.scala [hunk 16]
```diff
@@ -104,7 +102,7 @@ object Util extends LazyLogging {
       val transformations = dcInfo.key.transformations.toList
       infotonData => {
         val oldMeta = infotonData.meta
-        val newMeta = oldMeta.copy(path = transform(transformations, oldMeta.path))
+        val newPath = transform(transformations, oldMeta.path)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/Util.scala]: it's always baseinfoton.no inheritance between it so no need to match case

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/algo/AlgoFlow.scala [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+        new IndentityFlow().runAlgo(Map.empty[String, String])
+      case _ =>
+        val algoInfo = dcInfo.dcInfoExtra.get
+        val algoFlowInstance = URLClassLoader.loadClassFromJar[AlgoFlow](algoInfo.algoClass, algoInfo.algoJarUrl)
+        algoFlowInstance.runAlgo(algoInfo.algoParams)
+        }
+    Flow[InfotonData]
+      .map(_.data)
+      .via(algoFlow)
+      .map(rdf => InfotonData(InfotonThinMeta(rdf.utf8String.split(" ")(0)), rdf))
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/algo/AlgoFlow.scala]: done

File: server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+object ChildFirstURLClassLoader {
+  def loadClassFromJar[T](className: String, jarPath: String, excludes: Seq[String] = Seq()): T =
+    Loader(jarPath, excludes).load(className)
+
+  case class Loader(jarPath: String, excludes: Seq[String] = Seq()) {
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/ddpc/data/ChildFirstUrlClassLoader.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 19]
```diff
@@ -310,13 +316,14 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
 
   private def handleDcsToStart(dcsToStart: Seq[DcInfo]): Unit = {
     dcsToStart.foreach {
-      case dcInfo@DcInfo(dcKey, idxTimeFromUser, keyFromFinishedRun, tsvFile) => {
+      case dcInfo@DcInfo(dcKey, dcType, dcInfoExtra, idxTimeFromUser, keyFromFinishedRun, tsvFile) => {
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: remove dcType?

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 20]
```diff
@@ -310,13 +316,14 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
+//      case dcInfo@DcInfo(dcKey, idxTimeFromUser, keyFromFinishedRun, tsvFile) => {
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: remove

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 21]
```diff
@@ -325,33 +332,61 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
-                  logger.info(s"Starting sync for: $dcKey using position key $positionKey")
-                  self ! StartDcSync(dcInfo.copy(positionKey = Some(positionKey)))
-              }
           }
         }
       }
     }
   }
 
+  private def retievePositionFromZstoreAndStartDc(dcInfo: DcInfo, dcKey:DcInfoKey, idxTimeFromUser:Option[Long]) = {
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: duplicated arguments (dcInfo, dcKey)

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 22]
```diff
@@ -325,33 +332,61 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
+    val position = zStore.getStringOpt(dcInfo.dcInfoExtraType)
+    position.onComplete {
+      case Success(Some(zstorePosition)) =>
+        logger.info(s"Got key $zstorePosition from zstore for dc info key $dcKey")
+        self ! StartDcSync(dcInfo.copy(positionKey = Some(zstorePosition)))
+      case Success(None) => logger.info("first running")
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: each log print should contain the dcKey

Comment (e-orz) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: Also, rephrase the log

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 23]
```diff
@@ -325,33 +332,61 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
+        startDcFromIndexTime(dcInfo, dcKey, Future.successful(None))
+      case Failure(e) =>
+        logger.error(s"Sync ${dcInfo.key} failed with exception: ", e)
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: rephrase

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 24]
```diff
@@ -134,9 +130,15 @@ object InfotonRetriever extends LazyLogging {
                 val parsedResult: Try[RetrieveOutput] = Success(state._1.map {
                   //todo: validity checks that the data arrived correctly. e.g. that the path could be retrieved from _out etc.
                   im => {
-                    val parsed: (ByteStringBuilder, Option[Long]) =
-                      totals.parsed(im.meta.path)
-                    (InfotonData(im.meta, parsed._1.result), parsed._2)
+                    val parsed: (ByteStringBuilder, Option[Long]) = totals.parsed(im.base.path)
+                    val parsedData = if(parsed._2.isEmpty) {
+                      //uuids that don't have index time in cassandra. populate it using ES indexTime
+                      val idxTimeQuad = s"""${im.base.path} <cmwell://meta/sys#indexTime> "${im.indexTime}"^^<http://www.w3.org/2001/XMLSchema#long> ."""
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 25]
```diff
@@ -134,9 +130,15 @@ object InfotonRetriever extends LazyLogging {
...
+                      val q = "\"" + idxTimeQuad + "\""
+                      logger.warn(s"Sync $dcKey: Retrieve of uuid ${im.uuid.utf8String} didn't have index time. Adding $q from metadata manually")
+                      parsed._1.result ++ ByteString(idxTimeQuad)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala [hunk 26]
```diff
@@ -134,9 +130,15 @@ object InfotonRetriever extends LazyLogging {
...
+                    } else parsed._1.result
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonRetriever.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 27]
```diff
@@ -15,35 +15,29 @@
...
-import cmwell.dc.stream.akkautils.{ConcurrentFlow, DebugStage}
+import cmwell.dc.stream.akkautils.ConcurrentFlow
+import cmwell.dc.stream.algo.{AlgoFlow, DDPCAlgorithmJsonParser}
+import cmwell.dc.{LazyLogging, Settings}
+import cmwell.driver.Dao
 import cmwell.util.collections._
-import k.grid.{GridReceives, JvmIdentity}
-import play.api.libs.json._
+import cmwell.zstore.ZStore
+import com.typesafe.config.ConfigFactory
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 28]
```diff
@@ -218,6 +218,12 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
         //The sync was completed successfully. Save the position key for the next sync to start.
         val newSyncMap: SyncMap = currentSyncs - dcInfo.key + (dcInfo.key -> SyncerDone(dcInfo.positionKey.get))
         currentSyncs = newSyncMap
+        if(dcInfo.dcInfoExtraType != "remote")
+          zStore.putString(dcInfo.dcInfoExtraType , dcInfo.positionKey.get).onComplete{
+            case Success(res) =>
+            case Failure(ex) => logger.error(s"Failed to persist ${dcInfo.dcInfoExtraType } position", ex)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 19 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 20 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 21 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 22 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 23 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

**Pull request merged**